Library/Homebrew/extend/on_system.rb

```diff
@@ -26,25 +25,18 @@ def os_condition_met?(os_name, or_condition = nil)
       return true if [:macos, *MacOSVersions::SYMBOLS.keys].include?(os_name)
     end
 
-    if BASE_OS_OPTIONS.include?(os_name)
-      if Homebrew::SimulateSystem.none?
-        return OS.linux? if os_name == :linux
-        return OS.mac? if os_name == :macos
-      end
-
-      return Homebrew::SimulateSystem.send("#{os_name}?")
-    end
+    return Homebrew::SimulateSystem.send("treat_as_#{os_name}?") if BASE_OS_OPTIONS.include?(os_name)
 
     raise ArgumentError, "Invalid OS condition: #{os_name.inspect}" unless MacOSVersions::SYMBOLS.key?(os_name)
 
     if or_condition.present? && [:or_newer, :or_older].exclude?(or_condition)
       raise ArgumentError, "Invalid OS `or_*` condition: #{or_condition.inspect}"
     end
 
-    return false if Homebrew::SimulateSystem.linux? || (Homebrew::SimulateSystem.none? && OS.linux?)
+    return false if Homebrew::SimulateSystem.treat_as_linux?
```

Comment:
```suggestion
    return false if Homebrew::SimulateSystem.simulating_linux?
```

is another option

Reply:
I think I prefer the original because if we _aren't_ simulating Linux but are _actually_ on it, we still need to return true. I can see the word "simulate" being misleading

Reply:
If we do this:

```ruby
return false if Homebrew::SimulateSystem.os == :linux || OS.linux?
```

Then the condition will be true when we're simulating macOS on Linux (because `Homebrew::SimulateSystem.os` will be `:macos` (or a specific version) and `OS.linux?` will still be true). We only want this condition to be true when we're simulating Linux or we are on Linux _and not simulating anything_. So, the full check needs to be:

```ruby
return false if Homebrew::SimulateSystem.os == :linux || (Homebrew::SimulateSystem.os.blank? && OS.linux?)
```

The `Homebrew::SimulateSystem.treat_as_linux?` method is just a shortcut for this longer conditional.

Reply:
I feel like we want some sort of `OS.linux_no_really_not_simulated?`

Reply:
It's pretty hard to identify what cases we're trying to allow/not allow simulation 🤔

Reply:
So you're suggesting we just have `OS.linux?` and `OS.mac?` (and I guess by extension `Hardware::CPU.type`) defer to `Homebrew::SimulateSystem`? We can definitely do that. I was under the impression, though, that we really only wanted to scope the simulation code to generating JSON, and we wanted to avoid overriding `OS.linux?` and methods like that.

If we just want to have `OS` return stuff based on the simulated OS, should the logic for setting up the simulations just go in `OS` itself? So we can say `OS.simulate :linux` or something like that?

Reply:
I'm not sure, to be honest, I think other @Homebrew/maintainers could chime in here! I'm not sure where we are with `SimulateSystem` vs. `OS` vs. `HOMEBREW_SIMULATE_MACOS_ON_LINUX` 😅

Reply:
TBH, I don't see why we don't want `OS.mac?` and `OS.linux?` to lie when we ask it to (with `HOMEBREW_SIMULATE_MACOS_ON_LINUX`, or even `HOMEBREW_SIMULATE_LINUX_ON_MACOS`). But I'll admit to not having thought about this very hard or looked for corner cases.

I also don't really have a problem with `#simulating_linux?` being true even if the system really is running Linux, and it's not a simulation.

Are there cases where we want to distinguish between a machine that is running Linux with another machine running macOS, but the developer wants to trick `brew` into thinking its running Linux?

Reply:
> Are there cases where we want to distinguish between a machine that is running Linux with another machine running macOS, but the developer wants to trick `brew` into thinking its running Linux?

One instance might be that trying to `brew install` anything on macOS while simulating Linux might lead to a bad time. But do we really need guardrails here for users in this position? What if they actually really wanted to `brew install` some Linux software on macOS (perhaps for testing purposes)?

Reply:
> But do we really need guardrails here for users in this position?

Probably not...? The only way for a user to activate any simulation stuff is through `HOMEBREW_SIMULATE_MACOS_ON_LINUX`, and I _think_ it seems pretty clear that using that variable may cause problems if you don't know what you're doing. If we get weird bug reports where people are trying to set that unnecessarily, we can add a warning or undocumented it or something.

Reply:
In which case it seems to me that we should be fine with `#simulating_linux?`, no? Because we don't really need to distinguish between a real Linux machine and one only pretending to be Linux.

Reply:
> TBH, I don't see why we don't want OS.mac? and OS.linux? to lie when we ask it to (with HOMEBREW_SIMULATE_MACOS_ON_LINUX, or even HOMEBREW_SIMULATE_LINUX_ON_MACOS). But I'll admit to not having thought about this very hard or looked for corner cases.

I don't think they necessarily _should_ but the argument for them doing so is that it's useful to be able to e.g. generate information for macOS on Linux machines and, less so but still useful, to do Linux on macOS.

> Are there cases where we want to distinguish between a machine that is running Linux with another machine running macOS, but the developer wants to trick brew into thinking its running Linux?

I think the differentiation from me is "behaviour like we need to on Linux based on the underlying system" vs. "behaviour that differs on macOS on Linux and we want to generate data (e.g. dependency trees) on one OS from the other".

> One instance might be that trying to brew install anything on macOS while simulating Linux might lead to a bad time.

Yeh, specifically this: we shouldn't allow installing in this case. We can either e.g. prevent installation or opt-in to the commands/actions that we want to allow (e.g. `brew deps`, `brew info`, etc.)

> In which case it seems to me that we should be fine with `#simulating_linux?`, no?

I'm fine with behaviour, it's just the naming that's weird to me. `simulating_or_running_on_linux?` is a bit of a mouthful but more accurately describes what's going on. Alternatively, `SimulateSystem.linux?` is generic enough that it doesn't imply that you're _not_ on Linux right now.